dvc/stage/monitor.py

```diff
@@ -36,22 +37,141 @@ class MonitorTask:
+        super().__init__(stage, self._send_updates, proc)
+        import requests
+        from requests.adapters import HTTPAdapter, Retry
+
+        self.fs = stage.repo.fs
+        self.url = url
+        self.rev = stage.repo.scm.get_rev()
+
+        self.session = requests.Session()
+        self.session.headers.update({"content-type": "application/json"})
+        retries = Retry(
+            total=5,
+            # TODO: Is this enough for 429 errors?
+            backoff_factor=0.2,
+            # TODO: Which status to retry?
+            # 413 should Not be retried
+            status_forcelist=[429, 500, 502, 503, 504],
+        )
+
+        self.session.mount(
+            "https://studio.iterative.ai/", HTTPAdapter(max_retries=retries)
+        )
+
+        self.metrics_status, self.plots_status = self._collect_data()
+        self._mtime_metrics = {x: 0 for x in self.metrics_status}
+        self._mtime_plots = {x: 0 for x in self.plots_status}
+
+        self._last_plots_step = {x: -1 for x in self.plots_status}
+
+    def on_start(self):
+        self._send_updates()
+
+    def is_ready(self):
+        self._update_status()
+        return any(self.metrics_status.values()) or any(
+            self.plots_status.values()
+        )
+
+    def _collect_data(self):
+        metrics = {}
+        plots = {}
+        for out in self.stage.outs:
+            if out.is_metric:
+                metrics[out.def_path] = False
+            # Filter out non-linear plots and dirs, for MVP.
+            elif out.is_plot and not out.isdir():
+                plot_props = {} if isinstance(out.plot, bool) else out.plot
+                if plot_props.get("template", "linear") == "linear":
+                    plots[out.def_path] = False
+        return metrics, plots
+
+    def _update_status(self):
+        def _update_mtime(item, container):
+            if os.path.exists(item):
+                mtime = os.stat(item).st_mtime
+                if mtime != container[item]:
+                    container[item] = mtime
+                    return True
+            return False
+
+        for metric in self.metrics_status:
+            self.metrics_status[metric] = _update_mtime(
+                metric, self._mtime_metrics
+            )
+
+        for plot in self.plots_status:
+            self.plots_status[plot] = _update_mtime(plot, self._mtime_plots)
+
+    def _get_updates(self):
+        from dvc.repo.metrics.show import _read_metric
+        from dvc.repo.plots import parse
+
+        updates = {
+            # TODO: Is version enough?
+            # Should we include some more specific APPEND/REPLACE distinction?
+            "version": __version__,
+            "rev": self.rev,
+            "metrics": {},
+            "plots": {},
+        }
+        for metric, updated in self.metrics_status.items():
+            if updated:
+                updates["metrics"][metric] = _read_metric(
```

Comment:
Since you use `_read_metric()` you already sending errors on file level. :+1: `replace/append` should also operate at this level probably.